### PR TITLE
Assembly frames

### PR DESCRIPTION
**Pull request by unid15**
Missed get edition

Assemblies are still kind of a pain in the ass to work with, but now you can make more complex things with them.
- **Added assembly frames** - items that can store and link together an infinite amount of assemblies
- Using an assembly (timer, signaler, speaker, etc) on an assembly frame stores the assembly inside the frame
- Using the assembly frame's interface you can link different assemblies together
- If assembly A is linked to assembly B, whenever A sends a pulse, B will receive it. If B sends a pulse, A won't receive it (unless B is linked to A).
- There's no limit on the amount of links that an assembly can have
- Assembly frames can be built at the protolathe, or bought from vendomats for a coin
- **Added three 'assemblies' that are intended to be used in assembly frames. They can be made at the autolathe.
- **Added addition circuits** - whenever they receive a signal, their internal counter is increased by 1. Whenever their counter reaches a certain value, it's reset to 0 and the circuit emits a pulse.
- **Added randomizer circuits** - when pulsed, they randomly select some devices connected to them, and emit a pulse to them. It's possible to configure the amount of devices selected. They also generate a single random number from 0 to specified value
- **Added comparison circuits** - they must be used in conjunction with an addition circuit, randomizer circuit, timer or prox sensor, and another one/two assemblies. When pulsed, comparison circuits take the first connected assembly's value and check if it meets the condition (specified in the circuit). If it does, the second connected assembly is pulsed. Otherwise, the third connected assembly is pulsed. See the gallery for an example.
- Addition assemblies' value is their counter, randomizer assemblies' value is their last generated random number, timers and prox sensors have the time remaining (in secs) as their value
- All of the new assemblies can be made at the autolathe

**OTHER CHANGES**
- Timers and prox sensors can have a default time set up. When they tick down to 0, they'll reset to that default time instead of 0:10 like before.

- Timers and prox sensors now have a WIRE_RECEIVE wire. When they receive a pulse, they toggle their timer.
- Speakers can no longer have their message changed at distance

**GALLERY**

Here's my first machine; sending a signal to (activator) will turn it on/off. When turned on, it sings a daft punk song until turned off.

This machine's state can be toggled - depending on the state, either the first or the second circuit sends a pulse. Adding more addition circuits allows for more states.

This machine says a random number from 1 to 5 when activated.

This complex machine is operated by two buttons: one increases its internal counter by 1 (if the value goes over 5, it's reset to 0), second one checks if the counter is equal to 2 and plays an appropriate sound.

### CODE REVIEW

**Comment by unid15** [comment]
FUCK i missed the get

**Comment by ghost** [comment]
Would be great if you could create them with soldering iron, this opens a lot of possibilities.

**Comment by PJB3005** [comment]
This is very cool!

**Comment by Exxion** [comment]
ULTRA
FUCKING
SWEET

Although the first thing I'll probably use this for is making the radio completely unusable as antag

**Comment by unid15** [comment]
It's not like you already can't do that with a few proximity sensors

**Comment by Exxion** [comment]
I only wish voice sensors weren't coded to not listen to other assemblies or else it would be too easy

**Comment by FudgePucker** [comment]
Ideas.
Add the ability for AAC to send and receive signals, that way you can make an alert system that alerts you to atmospheric systems, or use its scripting system with the assembly frame to make even more complex computers.

Give mechanics better access to it and give them more goodies with it so the job is more unique. (See goons mechanics for inspiration, they have a crazy system)

File: code/modules/assembly/assembly_frame.dm [hunk 1]
```diff
@@ -0,0 +1,223 @@
...
+
+	icon_state = "assembly_box"
+	origin_tech = "programming=3;engineering=3;magnets=3"
+
+	var/list/obj/item/device/assembly/assemblies = list()
+
+	var/list/connections = list() //Assembly associated with the assembly it's connected to
+
+/obj/item/device/assembly_frame/Destroy()
+	connections.len = 0
```
Comment (PJB3005) [inline code/modules/assembly/assembly_frame.dm]: Has the `.len = 0` memory leak been fixed / proven to not be existant?

File: code/modules/assembly/logic.dm [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+#define sub_limit_href(amount) "<a href='?src=\ref[src];sub_limit=[amount]'>"
+
+/obj/item/device/assembly/addition/interact(mob/user as mob)
+	var/dat = ""
+
+	dat += "<tt>Addition circuit</tt><BR><BR>"
+	dat += "Counter: [sub_counter_href(10)]--</a> | [sub_counter_href(1)]-</a>  | <b>[pulse_counter]</b> | [add_counter_href(1)]+</a> | [add_counter_href(10)]++</a><BR>"
+	dat += "Limit: [sub_limit_href(10)]--</a> | [sub_limit_href(1)]-</a>  | <b>[limit]</b> | [add_limit_href(1)]+</a> | [add_limit_href(10)]++</a><BR>"
+
+	user << browse(dat, "window=circuit1")
```
Comment (PJB3005) [inline code/modules/assembly/logic.dm]: It'd be nice if you'd use browser datums for this.

**Comment by unid15** [comment]
Added two more fun assemblies to work with and put two assembly frames into each mechanic's locker

AAC integration is a neat idea but I still don't know anything about how AACs work, so maybe later. Now I'm gonna test everything again to ensure nothing is broken, and it'll be ready for merging

**Comment by unid15** [comment]
Tested and ready to go

**Comment by Probe1** [comment]
WIKI

UPDATE IT.

**Comment by Probe1** [comment]
Lynching averted.  Thank you for doing a good job with the wiki.

**Pull request merged**